salk_toolkit/io.py

```diff
@@ -189,13 +191,155 @@ def is_series_of_lists(s):
+        k (int, optional)           Number of new cols to create per subgroup.
+            Defaults to max and selects columns that include other vals than NA.
+        na_vals (list)              List of values to consider as NA.
+        scale (dict, optional)      Scale block to be added to each subgroup.
+        name (str, optional)        Name of the new columns. Defaults to
+            dict_with_create['name'] + "_topk".
+
+    Example:
+        df = pd.DataFrame({
+            'A_a_1': ['Mentioned'],
+            'A_a_2': ['Not mentioned'],
+            'B_b_1': ['Not mentioned'],
+            'B_b_2': ['Mentioned']
+        }, columns = ['A_a_1', 'A_a_2', 'B_b_1', 'B_b_2'])
+        meta = {
+            'from_columns': r'(.)_(.+)_(\d+)',
+            'res_cols': r'\1_\2_R\3',
+            'name': 'R10',
+            'k': 1
+        }
+        Creates subgroup ['A_a_1', 'A_a_2'] that map to ['A_a_R1'] 
+            and subgroup ['B_b_1', 'B_b_2'] that map to ['B_b_R1']
+            and creates metas with name 'R10_A_a' and 'R10_B_b'.
+        res = pd.DataFrame({
+            'A_a_R1': [1], #if scale has translate, 1 will be translated
+            'B_b_R1': [2]  #if scale has translate, 2 will be translated
+        }, columns = ['A_a_R1', 'B_b_R1'])
+        meta1 = {
+            'name': 'R10_A_a',
+            'columns': [['A_a_R1']]
+        }
+        meta2 = {
+            'name': 'R10_B_b',
+            'columns': [['B_b_R1']]
+        }
+    """
+    create = dict_with_create['create']
+    name = create.get('name', f"{dict_with_create['name']}_{create['type']}")
+    # Compile regex to catch df.columns that we want to apply topk for.
+    regex_from = re.compile(create['from_columns'])
+    from_cols = list(filter(lambda s: regex_from.match(s), df.columns))
+    # agg_ind is index of the regex group that we want to aggregate over.
+    # All other indeces are unique identifiers for each subgroup
+    # e.g. ['A', 'a'] and ['B', 'b'] (that are also added to column names)
+    # Recall that regex group 0 is the whole match.
+    # Note that re.Match.groups() does not include the whole match.
+    # This means that we need to subtract 1 from agg_ind.
+    agg_ind = create.get('agg_index', -1)
+    agg_ind = agg_ind-1 if agg_ind > 0 else agg_ind
+    n_groups = len(regex_from.match(from_cols[0]).groups())
+    has_subgroups = n_groups >= 2 # Multiple aggregations needed?
+    regex_to = create.get('res_cols', '')
+    kmax = create.get('k',None)
+    na_vals = create.get('na_vals', [])
+    if has_subgroups:
+        # collect all subgroups, later aggregate each subgroup separately
+        def get_subgroup_id(column):
+            "Discard the aggregation index and collect the rest as identifier."
+            subgroup_id = list(regex_from.match(column).groups())
+            subgroup_id.pop(agg_ind)
+            return subgroup_id
+        subgroups_ids = np.unique(
```

Comment:
I would have just done a set of tuples instead of bringing in numpy. But name is descriptive enough it does not matter much :)